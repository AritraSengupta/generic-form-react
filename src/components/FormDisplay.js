import React from 'react'
import PropTypes from 'prop-types'
import { Form, Grid, Message } from 'semantic-ui-react'
import { debounce, uniqBy, isEqual } from 'lodash'

import DatePicker from './Calendar/DatePicker'
import {
  getErrorList,
  getFormGridStruct,
  getValidatorMap,
  getDefaultState,
  convertToArray,
  runValidation,
  isRequired
} from '../utils'

export default class FormDisplay extends React.Component {
  constructor(props) {
    super(props)
    this.state = getDefaultState(props)
    this.handleDynamicSearchChange = debounce(
      this.handleDynamicSearchChange.bind(this),
      500
    )
    this.getFormUnit = this.getFormUnit.bind(this)
    this.updateStateIfPropsChange = this.updateStateIfPropsChange.bind(this)
    props.onChange(this.state)
  }

  componentDidUpdate(prevProps, prevState) {
    this.updateStateIfPropsChange(prevProps)
    if (!isEqual(prevState, this.state)) {
      const { onChange } = this.props
      onChange(this.state)
    }
  }

  updateStateIfPropsChange(prevProps) {
    if (!isEqual(prevProps, this.props)) {
      const newState = getDefaultState(this.props)
      this.setState(newState)
    }
  }

  /**
   * @values an array or a string representing the values that need validations
   * @validators an array of validators or an object with type and params or a custom funtion
   */
  validateInput = (values, validators, prevState, name, fieldname) => {
    if (!validators || (Array.isArray(validators) && !validators.length))
      return prevState.errors
    return {
      ...prevState.errors,
      [name]: runValidation(values, validators, fieldname)
    }
  }

  handleInputChange = (e, { name, value, validators, fieldname }) =>
    this.setState((prevState) => ({
      formData: { ...prevState.formData, [name]: value },
      errors: this.validateInput(value, validators, prevState, name, fieldname)
    }))

  handleCheckChange = (e, { name, value, checked, validators, fieldname }) =>
    this.setState((prevState) => {
      const currValue = { ...prevState.formData[name].value, [value]: checked }
      const selected = convertToArray(currValue)
      return {
        formData: {
          ...prevState.formData,
          [name]: {
            value: currValue,
            selected
          }
        },
        errors: this.validateInput(
          (selected || []).filter((s) => !!s),
          validators,
          prevState,
          name,
          fieldname
        )
      }
    })

  handleDropdownChange = (
    e,
    { name, value, validators, multiple, options, fieldname }
  ) => {
    let selectedOptions
    let values
    if (multiple) {
      selectedOptions = options.filter((o) => value.includes(o.value))
      values = selectedOptions.map((so) => so && so.text)
    } else {
      // eslint-disable-next-line prefer-destructuring
      selectedOptions = options.filter((o) => o.value === value)[0]
      values = selectedOptions && selectedOptions.text
    }
    this.setState((prevState) => ({
      formData: {
        ...prevState.formData,
        [name]: { value, selected: selectedOptions }
      },
      errors: this.validateInput(values, validators, prevState, name, fieldname)
    }))
  }

  addNewDropdownItem = (
    e,
    { name, value, multiple, validators, fieldname }
  ) => {
    if (multiple) {
      throw new Error(
        'Adding new elements with multiple dropdown not supported yet'
      )
    }
    const { dynamicOptions } = this.state
    const currentOption = { text: value, value: -1 }
    dynamicOptions[name].options = [
      ...dynamicOptions[name].options,
      currentOption
    ]

    this.setState((prevState) => ({
      formData: {
        ...prevState.formData,
        [name]: { value: currentOption.value, selected: currentOption }
      },
      errors: this.validateInput(value, validators, prevState, name, fieldname),
      dynamicOptions
    }))
  }

  async handleDynamicSearchChange(e, { name, multiple, searchQuery }) {
    if (!searchQuery.length) return
    const { dynamicOptions, formData } = this.state
    dynamicOptions[name].loading = true
    this.setState({ dynamicOptions })
    let newOptions
    try {
      newOptions = await dynamicOptions[name].callback(searchQuery)
    } catch {
      newOptions = []
    }
    const oldOptions = formData[name].selected || []
    if (multiple) {
      dynamicOptions[name].options = uniqBy(
        [...newOptions, ...oldOptions],
        'value'
      )
    } else {
      dynamicOptions[name].options = newOptions
    }
    dynamicOptions[name].loading = false
    this.setState({ dynamicOptions })
  }

  getFormUnit(val) {
    const { formData, errors, dynamicOptions } = this.state

    if (val.type === 'hidden' || (val.config && val.config.visible === false)) {
      return null
    }
    if (val.type === 'input') {
      return (
        <Form.Input
          key={val.dataId}
          required={isRequired(val.validators)}
          validators={val.validators}
          name={val.dataId}
          field_name={val.fieldname}
          label={val.fieldname}
          value={formData[val.dataId] || ''}
          error={!!(errors[val.dataId] && errors[val.dataId].length)}
          onChange={this.handleInputChange}
          disabled={val.config && val.config.disabled}
          {...((val.config && val.config.props) || {})}
          data-testid='input'
        />
      )
    }
    if (val.type === 'dropdown') {
      if (val.config && val.config.search) {
        const isMultiple = val.config && val.config.multiple
        let value = (formData[val.dataId] || {}).value
        if (!value && isMultiple) {
          // handling case for multiple dropdowns
          value = []
        }
        return (
          <Form.Dropdown
            key={val.dataId}
            label={val.fieldname}
            name={val.dataId}
            fieldname={val.fieldname}
            options={dynamicOptions[val.dataId].options}
            multiple={isMultiple}
            clearable={val.config && val.config.clearable}
            selection
            search
            placeholder={val.placeholder}
            value={value}
            onChange={this.handleDropdownChange}
            onSearchChange={this.handleDynamicSearchChange}
            disabled={
              (val.config && val.config.disabled) ||
              dynamicOptions[val.dataId].loading
            }
            loading={dynamicOptions[val.dataId].loading}
            onAddItem={this.addNewDropdownItem}
            allowAdditions={val.config && val.config.allowAdditions}
            validators={val.validators}
            error={!!(errors[val.dataId] && errors[val.dataId].length)}
            required={isRequired(val.validators)}
            data-testid='dropdown-search'
          />
        )
      }
      return (
        <Form.Dropdown
          key={val.dataId}
          label={val.fieldname}
          name={val.dataId}
          fieldname={val.fieldname}
          options={val.options}
          multiple={val.config && val.config.multiple}
          clearable={val.config && val.config.clearable}
          selection
          placeholder={val.placeholder}
          value={(formData[val.dataId] || {}).value}
          onChange={this.handleDropdownChange}
          disabled={val.config && val.config.disabled}
          onAddItem={this.addNewDropdownItem}
          allowAdditions={val.config && val.config.allowAdditions}
          loading={val.config && val.config.loading}
          validators={val.validators}
          error={!!(errors[val.dataId] && errors[val.dataId].length)}
          required={isRequired(val.validators)}
          data-testid='dropdown'
        />
      )
    }
    if (val.type === 'radio') {
      return (
        <React.Fragment key={val.dataId}>
          <Form.Field
            required={isRequired(val.validators)}
            error={!!(errors[val.dataId] && errors[val.dataId].length)}
          >
            <label>{val.fieldname}</label>
          </Form.Field>
          {val.options.map((radio) => (
            <Form.Radio
              key={radio.value}
              label={radio.text}
              name={val.dataId}
              fieldname={val.fieldname}
              value={radio.value}
              checked={formData[val.dataId] === radio.value}
              onChange={this.handleInputChange}
              validators={val.validators}
              error={!!(errors[val.dataId] && errors[val.dataId].length)}
            />
          ))}
        </React.Fragment>
      )
    }
    if (val.type === 'checkbox') {
      return (
        <React.Fragment key={val.dataId}>
          <Form.Field
            required={isRequired(val.validators)}
            error={!!(errors[val.dataId] && errors[val.dataId].length)}
          >
            <label>{val.fieldname}</label>
          </Form.Field>
          {val.options.map((check) => (
            <Form.Checkbox
              key={check.value}
              label={check.text}
              name={val.dataId}
              fieldname={val.fieldname}
              value={check.value}
              checked={
                !!((formData[val.dataId] || {}).value || {})[check.value]
              }
              onChange={this.handleCheckChange}
              validators={val.validators}
              error={!!(errors[val.dataId] && errors[val.dataId].length)}
            />
          ))}
        </React.Fragment>
      )
    }
    if (val.type === 'textarea') {
      return (
        <Form.TextArea
          key={val.dataId}
          required={isRequired(val.validators)}
          name={val.dataId}
          fieldname={val.fieldname}
          label={val.fieldname}
          value={formData[val.dataId] || ''}
          error={!!(errors[val.dataId] && errors[val.dataId].length)}
          onChange={this.handleInputChange}
          validators={val.validators}
        />
      )
    }
    if (val.type === 'date') {
      return (
        <React.Fragment key={val.dataId}>
          <Form.Field
            required={isRequired(val.validators)}
            error={!!(errors[val.dataId] && errors[val.dataId].length)}
          >
            <label>{val.fieldname}</label>
            <DatePicker
              data-testid='date'
              key={val.dataId}
              name={val.dataId}
              fieldname={val.fieldname}
              value={formData[val.dataId]}
              onDateChange={this.handleInputChange}
              validators={val.validators}
              {...((val.config && val.config.props) || {})}
            />
          </Form.Field>
        </React.Fragment>
      )
    }
    return <div>Improper Config</div>
  }

  validateForm = async () => {
    const { errors: currentErrors, formData: formState } = this.state
    const { data: form } = this.props
    // If error exists return
    if (Object.values(currentErrors).some((s) => s.length)) return true
    const validatorMap = getValidatorMap(form, formState)
    const errors = {}
    validatorMap.forEach((vm) => {
      if (vm.validators) {
        errors[vm.id] = runValidation(vm.value, vm.validators, vm.name)
      }
    })
    // We want to set the state first before making the final return of the function
    return new Promise((resolve) => {
      this.setState({ errors }, () => {
        const hasError = !!Object.values(errors).some((s) => s.length)
        resolve(hasError)
      })
    })
  }

  renderGrid(data, columns) {
    const gridList = getFormGridStruct(data, columns)
    const firstRowLength = (gridList[0] || []).length
    const actualColumns = columns < firstRowLength ? columns : firstRowLength
    return (
      <Grid columns={actualColumns} divided padded>
        {gridList.map((gl, idx) => (
          <Grid.Row key={`${gl.dataId}${idx}`}>
            {gl.map((gr, idx) => (
              <Grid.Column key={`${gr.dataId}${idx}`}>
                {this.getFormUnit(gr)}
              </Grid.Column>
            ))}
          </Grid.Row>
        ))}
      </Grid>
    )
  }

  render() {
    const { errors } = this.state
    const { data, loading, columns } = this.props

    const isColumnView = columns > 1

    return (
      <Form
        loading={!!loading}
        error={errors && Object.values(errors).some((s) => s.length)}
      >
        <Message
          error
          header='Please check the form for the following'
          list={getErrorList(errors)}
        />
        {isColumnView
          ? this.renderGrid(data, columns)
          : data.map((val) => this.getFormUnit(val))}
      </Form>
    )
  }
}

FormDisplay.propTypes = {
  data: PropTypes.arrayOf(
    PropTypes.exact({
      dataId: PropTypes.string.isRequired,
      type: PropTypes.oneOf([
        'hidden',
        'date',
        'input',
        'textarea',
        'dropdown',
        'checkbox',
        'radio'
      ]).isRequired,
      fieldname: PropTypes.string,
      defaultValue: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.number,
        PropTypes.object
      ]),
      config: PropTypes.object,
      options: PropTypes.arrayOf(
        PropTypes.shape({
          text: PropTypes.string,
          value: PropTypes.oneOfType([PropTypes.number, PropTypes.string])
        })
      ),
      validators: PropTypes.arrayOf(
        PropTypes.oneOfType([PropTypes.object, PropTypes.func])
      )
    })
  ).isRequired,
  onChange: PropTypes.func.isRequired,
  loading: PropTypes.bool,
  columns: PropTypes.number,
  errors: PropTypes.object
}
