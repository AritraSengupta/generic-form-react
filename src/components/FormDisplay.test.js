import React from 'react'
import { fireEvent, getAllByRole } from '@testing-library/react'
import FormDisplay from './FormDisplay'
import '@testing-library/jest-dom/extend-expect'

const data = [
  {
    dataId: 'id',
    type: 'hidden',
    defaultValue: null
  },
  {
    dataId: 'random',
    defaultValue: '',
    validators: [{ type: 'required' }]
  },
  {
    fieldname: 'Date',
    dataId: 'date',
    type: 'date',
    defaultValue: '1-1-20',
    validators: [{ type: 'required' }]
  },
  {
    fieldname: 'Name',
    dataId: 'name',
    type: 'input',
    defaultValue: 'abcd',
    validators: [{ type: 'required' }]
  },
  {
    fieldname: 'TextArea',
    dataId: 'textarea',
    type: 'textarea',
    defaultValue: '',
    validators: [{ type: 'required' }]
  },
  {
    fieldname: 'Dropdown',
    dataId: 'dropdown',
    type: 'dropdown',
    defaultValue: {},
    config: {
      clearable: true
    },
    options: [
      { text: 'Option1', value: 1 },
      { text: 'Option2', value: 2 }
    ],
    validators: [{ type: 'required' }]
  },
  {
    fieldname: 'Dropdown',
    dataId: 'dropdownMul',
    type: 'dropdown',
    defaultValue: {},
    config: {
      multiple: true,
      search: true,
      callback: jest.fn()
    },
    options: [
      { text: 'Option1', value: 1 },
      { text: 'Option2', value: 2 }
    ],
    validators: [{ type: 'required' }]
  },
  {
    fieldname: 'Checkbox',
    dataId: 'checkbox',
    type: 'checkbox',
    defaultValue: {},
    options: [
      { text: 'Option1', value: 1 },
      { text: 'Option2', value: 2 }
    ],
    validators: [{ type: 'required' }]
  },
  {
    fieldname: 'Radio',
    dataId: 'radio',
    type: 'radio',
    defaultValue: {},
    options: [
      { text: 'Option1', value: 1 },
      { text: 'Option2', value: 2 },
      { text: 'Option3', value: 3 }
    ],
    validators: [{ type: 'required' }]
  }
]

const formChangeMock = jest.fn()

describe('FormDisplay', () => {
  beforeEach(() => {
    formChangeMock.mockRestore()
  })

  it('should render', () => {
    const wrapper = render(<FormDisplay data={[]} onChange={() => {}} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('should render with all the units', async () => {
    const { getByTestId, container, rerender } = render(
      <FormDisplay data={data} onChange={formChangeMock} columns={2} />
    )
    const inputWrapper = getByTestId('input')
    expect(container).toMatchSnapshot()
    const inputElement = inputWrapper.children[0]
    expect(inputElement.value).toBe('abcd')
    fireEvent.change(inputElement, { target: { value: '1234' } })
    expect(inputElement.value).toBe('1234')
    rerender(<FormDisplay data={[]} onChange={formChangeMock} />)
  })

  it('should be able to change dropdown values', () => {
    const dropdownData = [
      {
        fieldname: 'Dropdown',
        dataId: 'dropdown',
        type: 'dropdown',
        defaultValue: { value: 1, selected: { text: 'Option1', value: 1 } },
        config: {
          clearable: true
        },
        options: [
          { text: 'Option1', value: 1 },
          { text: 'Option2', value: 2 }
        ],
        validators: [{ type: 'required' }]
      }
    ]
    const { getByTestId, container } = render(
      <FormDisplay data={dropdownData} onChange={formChangeMock} />
    )
    const dropdownWrapper = getByTestId('dropdown')
    expect(container).toMatchSnapshot()
    const dropdownElement = dropdownWrapper.children[0]
    const dropdownOptions = getAllByRole(dropdownWrapper, 'option')
    const dropdownOptionsText = dropdownOptions.map((d) => d.textContent)
    expect(dropdownElement.textContent).toBe(dropdownOptionsText[0])
    // Open the dropdown choices
    fireEvent.click(dropdownElement)
    // Click the next option
    fireEvent.click(dropdownOptions[1])
    expect(dropdownElement.textContent).toBe(dropdownOptionsText[1])
  })

  it('should be able to search for the search dropdown', () => {
    const dropdownData = [
      {
        fieldname: 'Dropdown',
        dataId: 'dropdown',
        type: 'dropdown',
        defaultValue: { value: 1, selected: { text: 'Option1', value: 1 } },
        config: {
          clearable: true,
          search: true,
          callback: () => {
            return [
              { text: 'Option3', value: 1 },
              { text: 'Option4', value: 2 }
            ]
          }
        },
        options: [
          { text: 'Option1', value: 1 },
          { text: 'Option2', value: 2 }
        ],
        validators: [{ type: 'required' }]
      }
    ]
    const { getByTestId, container } = render(
      <FormDisplay data={dropdownData} onChange={formChangeMock} />
    )
    const dropdownWrapper = getByTestId('dropdown-search')
    expect(container).toMatchSnapshot()
    const searchInput = dropdownWrapper.children[0]
    // Open the dropdown choices
    fireEvent.click(dropdownWrapper)
    fireEvent.click(searchInput)

    fireEvent.input(searchInput, { target: { value: 'Option' } })
    expect(dropdownWrapper).toMatchSnapshot()
    expect(searchInput.value).toBe('Option')
  })

  it('should validate form', () => {
    const dropdownData = [
      {
        fieldname: 'Dropdown',
        dataId: 'dropdown',
        type: 'dropdown',
        defaultValue: {},
        config: {
          clearable: true
        },
        options: [
          { text: 'Option1', value: 1 },
          { text: 'Option2', value: 2 }
        ],
        validators: [{ type: 'required' }]
      }
    ]
    const ref = React.createRef()
    const { container } = render(
      <FormDisplay data={dropdownData} onChange={formChangeMock} ref={ref} />
    )
    ref.current.validateForm()
    expect(container).toMatchSnapshot()
    expect(
      container.children[0].children[0].children[0].children[1].textContent
    ).toBe('Dropdown: is required')
  })
})
