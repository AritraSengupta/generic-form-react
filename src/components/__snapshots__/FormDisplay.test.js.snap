// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormDisplay should be able to change dropdown values 1`] = `
<div>
  <form
    class="ui form"
  >
    <div
      class="ui error message"
    >
      <div
        class="content"
      >
        <div
          class="header"
        >
          Please check the form for the following
        </div>
        <ul
          class="list"
        />
      </div>
    </div>
    <div
      class="required field"
    >
      <label>
        Dropdown
      </label>
      <div
        aria-expanded="false"
        class="ui selection dropdown"
        data-testid="dropdown"
        fieldname="Dropdown"
        name="dropdown"
        required=""
        role="listbox"
        tabindex="0"
        validators="[object Object]"
      >
        <div
          aria-atomic="true"
          aria-live="polite"
          class="divider text"
          role="alert"
        >
          Option1
        </div>
        <i
          aria-hidden="true"
          class="dropdown icon clear"
        />
        <div
          class="menu transition"
        >
          <div
            aria-checked="true"
            aria-selected="true"
            class="active selected item"
            role="option"
            style="pointer-events: all;"
          >
            <span
              class="text"
            >
              Option1
            </span>
          </div>
          <div
            aria-checked="false"
            aria-selected="false"
            class="item"
            role="option"
            style="pointer-events: all;"
          >
            <span
              class="text"
            >
              Option2
            </span>
          </div>
        </div>
      </div>
    </div>
  </form>
</div>
`;

exports[`FormDisplay should be able to search for the search dropdown 1`] = `
<div>
  <form
    class="ui form"
  >
    <div
      class="ui error message"
    >
      <div
        class="content"
      >
        <div
          class="header"
        >
          Please check the form for the following
        </div>
        <ul
          class="list"
        />
      </div>
    </div>
    <div
      class="required field"
    >
      <label>
        Dropdown
      </label>
      <div
        aria-busy="false"
        aria-disabled="false"
        aria-expanded="false"
        class="ui search selection dropdown"
        data-testid="dropdown-search"
        fieldname="Dropdown"
        name="dropdown"
        required=""
        role="combobox"
        validators="[object Object]"
      >
        <input
          aria-autocomplete="list"
          autocomplete="off"
          class="search"
          tabindex="0"
          type="text"
          value=""
        />
        <div
          aria-atomic="true"
          aria-live="polite"
          class="divider text"
          role="alert"
        >
          Option1
        </div>
        <i
          aria-hidden="true"
          class="dropdown icon clear"
        />
        <div
          class="menu transition"
          role="listbox"
        >
          <div
            aria-checked="true"
            aria-selected="true"
            class="active selected item"
            role="option"
            style="pointer-events: all;"
          >
            <span
              class="text"
            >
              Option1
            </span>
          </div>
          <div
            aria-checked="false"
            aria-selected="false"
            class="item"
            role="option"
            style="pointer-events: all;"
          >
            <span
              class="text"
            >
              Option2
            </span>
          </div>
        </div>
      </div>
    </div>
  </form>
</div>
`;

exports[`FormDisplay should be able to search for the search dropdown 2`] = `
<div
  aria-busy="false"
  aria-disabled="false"
  aria-expanded="true"
  class="ui active visible search selection dropdown"
  data-testid="dropdown-search"
  fieldname="Dropdown"
  name="dropdown"
  required=""
  role="combobox"
  validators="[object Object]"
>
  <input
    aria-autocomplete="list"
    autocomplete="off"
    class="search"
    tabindex="0"
    type="text"
    value="Option"
  />
  <div
    aria-atomic="true"
    aria-live="polite"
    class="divider text filtered"
    role="alert"
  >
    Option1
  </div>
  <i
    aria-hidden="true"
    class="dropdown icon clear"
  />
  <div
    class="visible menu transition"
    role="listbox"
  >
    <div
      aria-checked="true"
      aria-selected="true"
      class="active selected item"
      role="option"
      style="pointer-events: all;"
    >
      <span
        class="text"
      >
        Option1
      </span>
    </div>
    <div
      aria-checked="false"
      aria-selected="false"
      class="item"
      role="option"
      style="pointer-events: all;"
    >
      <span
        class="text"
      >
        Option2
      </span>
    </div>
  </div>
</div>
`;

exports[`FormDisplay should render 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class="ui form"
      >
        <div
          class="ui error message"
        >
          <div
            class="content"
          >
            <div
              class="header"
            >
              Please check the form for the following
            </div>
            <ul
              class="list"
            />
          </div>
        </div>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class="ui form"
    >
      <div
        class="ui error message"
      >
        <div
          class="content"
        >
          <div
            class="header"
          >
            Please check the form for the following
          </div>
          <ul
            class="list"
          />
        </div>
      </div>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`FormDisplay should render with all the units 1`] = `
<div>
  <form
    class="ui form"
  >
    <div
      class="ui error message"
    >
      <div
        class="content"
      >
        <div
          class="header"
        >
          Please check the form for the following
        </div>
        <ul
          class="list"
        />
      </div>
    </div>
    <div
      class="ui divided padded two column grid"
    >
      <div
        class="row"
      >
        <div
          class="column"
        >
          <div>
            Improper Config
          </div>
        </div>
        <div
          class="column"
        >
          <div
            class="required field"
          >
            <label>
              Date
            </label>
            <div
              class="ui fluid left icon input"
            >
              <input
                type="text"
                value="1-1-2020"
              />
              <i
                aria-hidden="true"
                class="calendar icon"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="column"
        >
          <div
            class="required field"
          >
            <label>
              Name
            </label>
            <div
              class="ui input"
              data-testid="input"
              field_name="Name"
              validators="[object Object]"
            >
              <input
                name="name"
                required=""
                type="text"
                value="abcd"
              />
            </div>
          </div>
        </div>
        <div
          class="column"
        >
          <div
            class="required field"
          >
            <label>
              TextArea
            </label>
            <textarea
              fieldname="TextArea"
              name="textarea"
              required=""
              rows="3"
              validators="[object Object]"
            />
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="column"
        >
          <div
            class="required field"
          >
            <label>
              Dropdown
            </label>
            <div
              aria-expanded="false"
              class="ui selection dropdown"
              data-testid="dropdown"
              fieldname="Dropdown"
              name="dropdown"
              required=""
              role="listbox"
              tabindex="0"
              validators="[object Object]"
            >
              <i
                aria-hidden="true"
                class="dropdown icon"
              />
              <div
                class="menu transition"
              >
                <div
                  aria-checked="false"
                  aria-selected="true"
                  class="selected item"
                  role="option"
                  style="pointer-events: all;"
                >
                  <span
                    class="text"
                  >
                    Option1
                  </span>
                </div>
                <div
                  aria-checked="false"
                  aria-selected="false"
                  class="item"
                  role="option"
                  style="pointer-events: all;"
                >
                  <span
                    class="text"
                  >
                    Option2
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="column"
        >
          <div
            class="required field"
          >
            <label>
              Dropdown
            </label>
            <div
              aria-busy="false"
              aria-disabled="false"
              aria-expanded="false"
              class="ui multiple search selection dropdown"
              data-testid="dropdown-search"
              fieldname="Dropdown"
              name="dropdownMul"
              required=""
              role="combobox"
              validators="[object Object]"
            >
              <input
                aria-autocomplete="list"
                autocomplete="off"
                class="search"
                tabindex="0"
                type="text"
                value=""
              />
              <span
                class="sizer"
              />
              <i
                aria-hidden="true"
                class="dropdown icon"
              />
              <div
                aria-multiselectable="true"
                class="menu transition"
                role="listbox"
              >
                <div
                  aria-checked="false"
                  aria-selected="true"
                  class="selected item"
                  role="option"
                  style="pointer-events: all;"
                >
                  <span
                    class="text"
                  >
                    Option1
                  </span>
                </div>
                <div
                  aria-checked="false"
                  aria-selected="false"
                  class="item"
                  role="option"
                  style="pointer-events: all;"
                >
                  <span
                    class="text"
                  >
                    Option2
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="column"
        >
          <div
            class="required field"
          >
            <label>
              Checkbox
            </label>
          </div>
          <div
            class="field"
          >
            <div
              class="ui checkbox"
              fieldname="Checkbox"
              validators="[object Object]"
            >
              <input
                class="hidden"
                name="checkbox"
                readonly=""
                tabindex="0"
                type="checkbox"
                value="1"
              />
              <label>
                Option1
              </label>
            </div>
          </div>
          <div
            class="field"
          >
            <div
              class="ui checkbox"
              fieldname="Checkbox"
              validators="[object Object]"
            >
              <input
                class="hidden"
                name="checkbox"
                readonly=""
                tabindex="0"
                type="checkbox"
                value="2"
              />
              <label>
                Option2
              </label>
            </div>
          </div>
        </div>
        <div
          class="column"
        >
          <div
            class="required field"
          >
            <label>
              Radio
            </label>
          </div>
          <div
            class="field"
          >
            <div
              class="ui radio checkbox"
              fieldname="Radio"
              validators="[object Object]"
            >
              <input
                class="hidden"
                name="radio"
                readonly=""
                tabindex="0"
                type="radio"
                value="1"
              />
              <label>
                Option1
              </label>
            </div>
          </div>
          <div
            class="field"
          >
            <div
              class="ui radio checkbox"
              fieldname="Radio"
              validators="[object Object]"
            >
              <input
                class="hidden"
                name="radio"
                readonly=""
                tabindex="0"
                type="radio"
                value="2"
              />
              <label>
                Option2
              </label>
            </div>
          </div>
          <div
            class="field"
          >
            <div
              class="ui radio checkbox"
              fieldname="Radio"
              validators="[object Object]"
            >
              <input
                class="hidden"
                name="radio"
                readonly=""
                tabindex="0"
                type="radio"
                value="3"
              />
              <label>
                Option3
              </label>
            </div>
          </div>
        </div>
      </div>
    </div>
  </form>
</div>
`;

exports[`FormDisplay should validate form 1`] = `
<div>
  <form
    class="ui error form"
  >
    <div
      class="ui error message"
    >
      <div
        class="content"
      >
        <div
          class="header"
        >
          Please check the form for the following
        </div>
        <ul
          class="list"
        >
          <li
            class="content"
          >
            Dropdown: is required
          </li>
        </ul>
      </div>
    </div>
    <div
      class="error required field"
    >
      <label>
        Dropdown
      </label>
      <div
        aria-expanded="false"
        aria-invalid="true"
        class="ui selection dropdown"
        data-testid="dropdown"
        fieldname="Dropdown"
        name="dropdown"
        required=""
        role="listbox"
        tabindex="0"
        validators="[object Object]"
      >
        <i
          aria-hidden="true"
          class="dropdown icon"
        />
        <div
          class="menu transition"
        >
          <div
            aria-checked="false"
            aria-selected="true"
            class="selected item"
            role="option"
            style="pointer-events: all;"
          >
            <span
              class="text"
            >
              Option1
            </span>
          </div>
          <div
            aria-checked="false"
            aria-selected="false"
            class="item"
            role="option"
            style="pointer-events: all;"
          >
            <span
              class="text"
            >
              Option2
            </span>
          </div>
        </div>
      </div>
    </div>
  </form>
</div>
`;
